### ISSUE TITLE
poll based encrypt/decrypt/sign api to support both async and sync call

### ISSUE THREAD

**Issue by taikulawo** [rustls/rustls#1976]
**Checklist**
* [x] I've searched the issue tracker for similar requests
https://github.com/rustls/rustls/issues/850#issuecomment-1544457213
https://github.com/rustls/rustls/issues/341
https://github.com/rustls/rustls/pull/1648
https://github.com/rustls/rustls/issues/1808
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**
https://docs.rs/rustls/latest/rustls/crypto/cipher/trait.MessageDecrypter.html

convert crypto api to poll based

from
```rs
pub trait MessageDecrypter: Send + Sync {
    /// Decrypt the given TLS message `msg`, using the sequence number
    /// `seq` which can be used to derive a unique [`Nonce`].
    fn decrypt<'a>(
        &mut self,
        msg: InboundOpaqueMessage<'a>,
        seq: u64,
    ) -> Result<InboundPlainMessage<'a>, Error>;
}
```
to
```rs
pub trait MessageDecrypter: Send + Sync {
    /// Decrypt the given TLS message `msg`, using the sequence number
    /// `seq` which can be used to derive a unique [`Nonce`].
    fn decrypt<'a>(
        &mut self,
        cx: &mut Context<'a>,
        msg: InboundOpaqueMessage<'a>,
        seq: u64,
    ) -> Result<InboundPlainMessage<'a>, Error>;
}
```
so that we can call `tokio#poll_read` to make encrypt/decrypt async

every encrypt/decrypt should call through trait, so we can hack it

**Additional context**
rustls only support poll method. user decide whether sync or async. If sync, just do and return `Poll::Ready`

**Comment by ctz** [comment]
I think It's exceedingly unlikely we'll do this. `decrypt` (and `encrypt`) are computation-bound and relatively small computations at that, and not IO-bound. What is your use case?

**Comment by djc** [comment]
Have you looked at tokio-rustls and the [unbuffered](https://docs.rs/rustls/latest/rustls/unbuffered/index.html) API?

**Comment by taikulawo** [comment]
> I think It's exceedingly unlikely we'll do this. `decrypt` (and `encrypt`) are computation-bound and relatively small computations at that, and not IO-bound. What is your use case?

I want to hook asymmetric cryptography sign and encrypt/decrypt phase, send material to hardware cluster over internet then computed and send back. 

It like cloudflare keyless using intel QAT over internet, so rustls encrypt/decrypt must async

**Comment by taikulawo** [comment]
> Have you looked at tokio-rustls and the [unbuffered](https://docs.rs/rustls/latest/rustls/unbuffered/index.html) API?

I see. I can parse record to extract key material and compute on my hardware cluster through internet. I will do asymmetric cryptography, but I can't let rustls skip from handshake phase to application data phase (symmetric cryptography)

**Comment by taikulawo** [comment]
> > It like cloudflare keyless
> 
> I think this is already covered in #850 ("asynchronous handshake authentication signature" cases)

yes, partially.

Does there have async version `MessageDecrypter` and `MessageEncrypter`？

**Comment by taikulawo** [comment]
Our company has deeply demands on async sign feature to offload asymmetric key exchange to hardware accelerator. so I would like to see this feature shipped into rustls.

Which async pattern you would like to accepted? poll-based, future, async trait fn or others. Or what async design you want to see in rustls?

I will submit a PR. @djc @ctz

**Comment by fangkaiwodexigua** [comment]
we also need it. asynchornous tls sign

**Comment by venkatesh6911** [comment]
Following up on #2762, which was merged into this issue, I've been prototyping the concrete design this issue is asking for, scoped to `Signer`/`SigningKey`.

Motivation is the same hardware-offload case raised here and in #2762: OpenSSL engines/providers that use the `ASYNC_JOB` API (e.g. Intel QAT) to hand asymmetric operations to a hardware/async execution context, where blocking the handshake thread for the round-trip defeats the point of an event-driven server.

Rough shape of what I've been experimenting with:
- `Signer::sign()` can report that it needs to pause instead of only returning `Result<Vec<u8>, Error>` immediately, via a small pending-operation type.
- The connection exposes something like `has_pending_operation()`/`pending_operation_fd()`, so callers can wait on a real fd/waker instead of blocking the calling thread.
- This is deliberately scoped to signing trait and is meant to complement the externalized-state-machine direction in #2397 and the recently-merged #3140 .

@ctz, re your note on #2762 that ASYNC/executor integration "isn't going to happen within the context of this project" - I read that as specifically about `rustls-openssl-compat` acting as an executor bridge, not about `rustls` core exposing a non-blocking `Signer` hook that any executor could drive itself. Wanted to check that reading is correct before proceeding further. Is this still the right issue for that, or would you rather it move elsewhere given #2397 is still unresolved?

I can raise a draft PR for you to take a look at and give feedback on.
